### DISCUSSION TITLE
support pci device hotplug on guest kernel version < 5.5

### DISCUSSION THREAD

**Discussion by Y-hwan** [cloud-hypervisor/cloud-hypervisor#5028] [Q&A] [answered]
I want to implement some functionality in cloud-hypervisor
1. support pci device hotplug on guest kernel version < 5.5, which is limited by GED because cloud-hypervisor use HW-reduced ACPI platforms. 
2. support legacy bios.

I looked up the ACPI spec, HW-reduced ACPI doesn't seems to support legacy bios. Then I came up with two solutions.
1. Reserving HW-reduced ACPI.Use GPIO hardware to support pci device hotplug and UEFI with CSM to support legacy bios. It might seem simpler.
2.Use Fixed Hardware. I haven't figured out what else I need to do if choosing this.

Because I'm unacquainted with ACPI,  I'm not sure about the correctness of the above.
Could anyone give me some ideas?

**Comment by rbradford** [comment, marked as answer]
We made an intentional decision to only support the ACPI HW-Reduced platform as it is significantly simpler and involves less legacy functionality. The 5.5 kernel was released almost three years ago and there are plenty of OS versions that include it or newer kernels. Unfortunately if you must run an older kernel then your option is either to forgo guest device hotplug or backport the fix (https://github.com/torvalds/linux/commit/ac36d37e943635fc072e9d4f47e40a48fbcdb3f0). I don't know what you mean by "legacy bios" in terms of hotplug as it has nothing to do with that.
